File: lib/src/data/loader_yaml.ts
import { Account, Type as AccountType } from '../core/account';
import { Balance, Type as BalanceType } from '../core/balance';
import { Month } from '../core/month';
import { BudgetBuilder, TransferData } from '../core/budget';

import * as yaml from 'js-yaml';

interface BalanceData {
  grp?: string;
  date?: Date;
  camt?: number;
  pamt?: number;
}

interface TransferYamlData {
  grp?: string;
  date?: Date;
  camt?: number;
  pamt?: number;
  desc?: string;
}

export interface YamlData {
  name?: string;
  desc?: string;
  number?: string;
  type?: string;
  opened_on?: string;
  closed_on?: string;
  owner?: string[];
  url?: string;
  phone?: string;
  address?: string;
  username?: string;
  pswd?: string;
  balances?: BalanceData[];
  transfers_to?: { [key: string]: TransferYamlData[] };
}

function lookupAccountType(type: string): AccountType | undefined {
  for (const [key, value] of Object.entries(AccountType)) {
    if (value == type) {
      return AccountType[key as keyof typeof AccountType];
    }
  }
  return undefined;
}

function isValidDate(date: any) {
  return date && Object.prototype.toString.call(date) === '[object Date]' && !isNaN(date);
}

function makeBalance(data: BalanceData) {
  let amount = 0;
  let balanceType = BalanceType.UNKNOWN;
  if (data.camt !== undefined) {
    amount = Math.round(100 * data.camt);
    balanceType = BalanceType.CONFIRMED;
  } else if (data.pamt !== undefined) {
    amount = Math.round(100 * data.pamt);
    balanceType = BalanceType.PROJECTED;
  } else {
    throw new Error(
      `Balance ${JSON.stringify(data)} does not have amount type set, expected camt or pamt entry.`
    );
  }
  if (!data.date || !isValidDate(data.date)) {
    throw new Error(`Balance ${JSON.stringify(data)} does not have date set.`);
  }
  return new Balance(amount, data.date, balanceType);
}

function processYamlData(budgetBuilder: BudgetBuilder, data: YamlData) {
  if (!data.name) {
    // Ignore data which dont represent account.
    return;
  }
  if (!data.type) {
    throw new Error(`Type is not set for account '${data.name}'`);
  }
  const accountType = lookupAccountType(data.type);
  if (!accountType) {
    throw new Error(`Unknown type '${data.type}' for account '${data.name}'`);
  }
  if (!data.owner || data.owner.length === 0) {
    throw new Error(`Account '${data.name}' has no owners`);
  }
  const account = new Account({
    name: data.name,
    description: data.desc,
    type: accountType,
    number: data.number,
    openedOn: data.opened_on ? Month.fromString(data.opened_on) : undefined,
    closedOn: data.closed_on ? Month.fromString(data.closed_on) : undefined,
    owners: data.owner,
    url: data.url,
    phone: data.phone,
    address: data.address,
    userName: data.username,
    password: data.pswd,
  });
  budgetBuilder.setAccount(account);
  if (data.balances) {
    for (const balanceData of data.balances) {
      if (!balanceData.grp) {
        throw Error(`Balance entry ${JSON.stringify(balanceData)} has no grp setting.`);
      }
      let month;
      try {
        month = Month.fromString(balanceData.grp);
      } catch (e) {
        const message = e instanceof Error ? e.message : 'unknown';
        throw new Error(`Balance ${JSON.stringify(balanceData)} has bad grp setting: ${message}`);
      }
      const balance = makeBalance(balanceData);
      const balanceMonthDiff = Math.abs(
        balance.date.getUTCFullYear() * 12 +
          balance.date.getUTCMonth() -
          month.year * 12 -
          month.month
      );
      if (balanceMonthDiff > 2) {
        throw new Error(
          `For ${account.name} account ${balance} and month ${month} are ${balanceMonthDiff} months apart (2 max).`
        );
      }
      budgetBuilder.setBalance(account.name, month.toString(), balance);
    }
  }
  if (data.transfers_to) {
    for (const [account_name, transfers] of Object.entries(data.transfers_to)) {
      if (!transfers) continue;
      for (const transfer_data of transfers) {
        if (!transfer_data.grp) {
          throw new Error(
            `For account "${account.name} transfer to ${account_name}" does not have "grp" field.`
          );
        }
        if (!transfer_data.date || !isValidDate(transfer_data.date)) {
          throw new Error(
            `For account "${account.name}" transfer to "${account_name}" does not have a valid "date" field.`
          );
        }
        let balance: Balance | undefined;
        if (transfer_data.pamt !== undefined) {
          balance = new Balance(
            Math.round(100 * transfer_data.pamt),
            transfer_data.date,
            BalanceType.PROJECTED
          );
        } else if (transfer_data.camt !== undefined) {
          balance = new Balance(
            Math.round(100 * transfer_data.camt),
            transfer_data.date,
            BalanceType.CONFIRMED
          );
        }
        if (balance === undefined) {
          throw new Error(
            `For account "${account.name}" transfer to "${account_name}" ` +
              `does not have "pamt" or "camt" field: ${JSON.stringify(transfer_data)}.`
          );
        }

        const transferMonth = Month.fromString(transfer_data.grp);
        const balanceMonth = Month.fromDate(balance.date);
        if (Math.abs(balanceMonth.distance(transferMonth)) > 2) {
          throw new Error(
            `For account "${account.name}" transfer to "${account_name}" ` +
              `for ${transferMonth} date ${balance.date
                .toISOString()
                .slice(0, 10)} (${balanceMonth}) are too far apart.`
          );
        }
        const transfer: TransferData = {
          fromAccount: account.name,
          fromMonth: transferMonth,
          toAccount: account_name,
          toMonth: transferMonth,
          balance,
          description: transfer_data.desc,
        };

        budgetBuilder.addTransfer(transfer);
      }
    }
  }
}

export function parseYamlContent(content: string, relativeFilePath: string): YamlData | undefined {
  return yaml.load(content, {
    filename: relativeFilePath,
    onWarning: (ex: yaml.YAMLException) => {
      console.warn(ex);
    },
    schema: yaml.DEFAULT_SCHEMA,
    json: false,
  }) as YamlData | undefined;
}

export function loadYamlFile(
  budgetBuilder: BudgetBuilder,
  accountData: YamlData,
  relative_file_path: string
): void {
  try {
    processYamlData(budgetBuilder, accountData);
  } catch (e) {
    const message = ' while processing ' + relative_file_path;
    if (e instanceof Error) {
      e.message += message;
      console.error(e.name + ': ' + e.message);
    } else {
      console.error('Error' + message);
    }
    console.log('Account Data', accountData);
    throw e;
  }
}

File: server/kt/src/main/kotlin/com/cherba29/tally/data/LoaderYaml.kt
package com.cherba29.tally.data

import com.cherba29.tally.core.Account
import com.cherba29.tally.core.Balance
import com.cherba29.tally.core.Month
import com.cherba29.tally.data.builder.BudgetBuilder
import com.cherba29.tally.data.yaml.BalanceYamlData
import com.cherba29.tally.data.yaml.YamlData
import io.github.oshai.kotlinlogging.KotlinLogging
import java.nio.file.Path
import kotlin.math.abs
import kotlin.math.roundToLong

private fun YamlData.toAccount(): Account? {
  if (name == null) {
    // Ignore data which dont represent account.
    return null
  }
  if (path == null) {
    logger.warn { "Account '$name' has no path set" }
  }
  if (owner.isNullOrEmpty()) {
    throw IllegalArgumentException("Account '$name' has no owners")
  }
  if (desc == null) {
    logger.warn { "$name is missing description field." }
  }
  if (path.isNullOrEmpty()) {
    throw IllegalArgumentException("$name is missing path field")
  }
  if (openedOn == null) {
    throw IllegalArgumentException("$name is missing opened_on field")
  }

  return Account(
    name = name,
    path = path,
    owners = owner.toSet(),
    rank = rank,
    description = desc,
    number = number,
    openedOn = openedOn,
    closedOn = closedOn,
    url = url,
    phone = phone,
    address = address,
    userName = username,
    password = pswd,
  )
}

private fun BalanceYamlData.toBalance(name: String): Balance {
  requireNotNull(grp)
  var amount: Long
  var balanceType: Balance.Type
  if (camt != null) {
    amount = (100.0 * camt).roundToLong()
    balanceType = Balance.Type.CONFIRMED
  } else if (pamt != null) {
    amount = (100.0 * pamt).roundToLong()
    balanceType = Balance.Type.PROJECTED
  } else {
    throw IllegalArgumentException("For $name account balance for $grp does not have amount type set, expected camt or pamt entry")
  }
  if (date == null) {
    throw IllegalArgumentException("For $name account balance for $grp does not have date set")
  }
  val balance =  Balance(amount, date, balanceType, desc ?: "")
  val balanceMonthDiff = abs(balance.date.year * 12 + balance.date.month.ordinal - grp.year * 12 - grp.month)
  // TODO: move logical inconsistency checks into budget builder.
  if (balanceMonthDiff > 2) {
    throw IllegalArgumentException(
      "For $name account $balance and month $grp are $balanceMonthDiff months apart (2 max)"
    )
  }
  return balance
}

/**
 * Validates YamlData and atomically adds it to the budget.
 */
private fun processYamlData(budgetBuilder: BudgetBuilder, data: YamlData): Boolean {
  // Ignore data which dont represent account.
  val account = data.toAccount() ?: return false
  val convertedBalances = data.balances?.associate {
    (it.grp ?: throw IllegalArgumentException(
      "For ${account.name} account balance entry with date '${it.date}' and desc '${it.desc}' has no grp setting"
    )) to it.toBalance(account.name) } ?: mapOf()

  budgetBuilder.addAccount(account, convertedBalances)

  if (data.transfersTo != null) {
    for ((toAccountName, transfers) in data.transfersTo.entries) {
      if (transfers == null) continue
      for (transferData in transfers) {
        if (transferData.grp == null) {
          throw IllegalArgumentException(
            "For account '${account.name}' transfer to '$toAccountName' does not have 'grp' field"
          )
        }
        if (transferData.date == null) {
          throw IllegalArgumentException(
            "For account '${account.name}' transfer to '$toAccountName' does not have a valid 'date' field"
          )
        }
        var balance: Balance? = null
        if (transferData.pamt != null) {
          balance = Balance(
            (100 * transferData.pamt).roundToLong(),
            transferData.date,
            Balance.Type.PROJECTED
          )
        } else if (transferData.camt != null) {
          balance = Balance(
            (100 * transferData.camt).roundToLong(),
            transferData.date,
            Balance.Type.CONFIRMED
          )
        }
        if (balance == null) {
          throw IllegalArgumentException(
            "For account '${account.name}' transfer to '$toAccountName' " +
                "for ${transferData.grp} ${transferData.date} " +
                "does not have 'pamt' or 'camt' field"
          )
        }

        // TODO: move logical inconsistency checks into budget builder.
        val transferMonth = transferData.grp
        val balanceMonth = Month.fromDate(balance.date)
        if (abs(balanceMonth - transferMonth) > 2) {
          throw IllegalArgumentException(
            "For account '${account.name}' transfer to '${toAccountName}' " +
                "for $transferMonth date ${balance.date} (${balanceMonth}) are too far apart"
          )
        }
        budgetBuilder.addAccountTransfer(
          account,
          toAccountName,
          transferMonth,
          balance,
          transferData.desc,
          transferData.tags ?: listOf()
        )
      }
    }
  }
  return true
}

fun loadYamlFile(budgetBuilder: BudgetBuilder, accountData: YamlData, relativeFilePath: Path): Boolean {
  return try {
    processYamlData(budgetBuilder, accountData)
  } catch (e: IllegalArgumentException) {
    logger.error { e.javaClass.simpleName + ": " + e.message }
    throw IllegalArgumentException(e.message + ", while processing $relativeFilePath", e)
  } catch (e: Exception) {
    val message = ", while processing $relativeFilePath"
    logger.error { e.javaClass.simpleName + ": " + e.message + message }
    logger.info { "Account Data$accountData" }
    throw e
  }
}

private val logger = KotlinLogging.logger {}
